### DISCUSSION TITLE
OpenCV and how to convert Image to Mat

### DISCUSSION THREAD

**Discussion by kt10aan** [robotology/community#322] [Q&A] [answered]
In some older code there was: (I am leaving out the if port open checks)

```cpp
bool YarpImageProvider::get(cv:Mat& frame) {
    yarp::sig::ImageOf<yarp::sig::PixelRgb> *image = imagePort.read(true);
    if (image) {
        cv::Mat rgbframe((IplImage*)image->getIplImage);
        cv::cvtColor(rgbframe,frame, CV_RGB2BGR);
        [...]
```

and was working fine until the last upgrade started a chain of dependencies update and now we are using OpenCV3. The above code was written for OpenCV 2.4.9, so quite ancient and doesn't accept any Ipl -based methods.

What I did is replacing line 4 with:
```cpp
cv::Mat rgbframe(image->height, image->width, 21, image);
```

However, when running the software is segfaults. Type 21 is ``CV_32FC3``. I tried with other image types but most of them give me the same segfault and a few other give me a garbled image.

My question is: what would be the correct line 4 to convert the YARP image to a ``cv::Mat``?

**Comment by claudiofantacci** [comment, marked as answer]
Hi @kt10aan, in OpenCV 3.* they removed the `cv::Mat `constructor with `IplImage`, i.e.
```cpp
Mat(const IplImage *img, bool copyData=false)
```
However, you can still use _almost_ the same line of code.
With the following two lines you can wrap a `yarp::sig::ImageOf<yarp::sig::PixelRgb>` with a `cv::Mat`, i.e. they share the same memory representing pixel values.

```cpp
yarp::sig::ImageOf<yarp::sig::PixelRgb> *yarp_img = imagePort.read(true);
cv::Mat img = cv::cvarrToMat(static_cast<IplImage*>(yarp_img->getIplImage()));
```

**Comment by kt10aan** [comment]
@claudiofantacci , many thanks! it worked! (but it is ''yarp_img->getIplImage'').
Again, many thanks:)

**Comment by pattacini** [comment]
See also #122.
